dbms/src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3255,11 +3255,26 @@ ReservationPtr MergeTreeData::tryReserveSpacePreferringTTLRules(UInt64 expected_
             reservation = destination_ptr->reserve(expected_size);
             if (reservation)
                 return reservation;
+            else
+                if (ttl_entry->destination_type == PartDestinationType::VOLUME)
+                    LOG_WARNING(log, "Would like to reserve space on volume '"
+                            << ttl_entry->destination_name << "' by TTL rule of table '"
+                            << log_name << "' but there is not enough space");
+                else if (ttl_entry->destination_type == PartDestinationType::DISK)
+                    LOG_WARNING(log, "Would like to reserve space on disk '"
+                            << ttl_entry->destination_name << "' by TTL rule of table '"
+                            << log_name << "' but there is not enough space");
         }
     }
 
     reservation = storage_policy->reserve(expected_size, min_volume_index);
 
+    if (!reservation && min_volume_index > 0)
+    {
+        LOG_WARNING(log, "Could not reserve " << expected_size << " bytes on volume with index starting from " << min_volume_index << ", falling back to volumes with lesser index");
```

Comment:
questionable.

Let's imagine we have 500 Gb SSD + 5 Tb HDD.
HDD is almost full. Some huge merge assigned (by leader replica). If that merge will not be executed - it's not very good, but nothing bad happen. It will send exceptions in logs, some metrics will sign that something doesn't work properly, somebody will need to make maintainance tasks to cleanup hdd. But inserts are (relatively) safe, as SSD has lot of space.

If we will make that huge merge to SSD it will most probably lead soon to out-of-space on SSD too, and inserts will stop working (what is significantly worse, as can lead to data loss).

Reply:
So basically we need to make selection logic more complicated in order to not allow such merges at all.

Reply:
You can't avoid them at all (they can be assigned by another replica which has plenty of disk space).

Reply:
Doesn't it worth making that fix for non-replicated tables?

Reply:
Keeping enough free space on each particular disk is a normal maintenance operation. It should not happen when some disk go out of disk space. But if it happens I think we should not start to propagate the problem to others (most probably much smaller) disks. 

I.e. let's say I planned to store 7 days of data on SSD, and the rest on HDD.
And I know that 7 days of data will take about (for example) 350 Gb.

If I forget to setup a cleanup on HDD and after few months my HDD will be full, i want to have that problem only on HDD. Not on SSD.

Reply:
I just wonder where shall downloaded part go? I mean on which volume? It has no connection with subparts and therefore can be allocated on any disk. And that will basically mean exactly this change under discussion but even worse, because it will scan volumes starting from index 0.

Reply:
We need to pass TTL information in part header when downloading. Similarly, passing of the part size was added.

Reply:
See https://github.com/ClickHouse/ClickHouse/blob/8fb95419584016610de5b2ce2a31f76314568743/dbms/src/Storages/MergeTree/DataPartsExchange.cpp#L116

Reply:
I guess this needs a little bit more investigation. Which information shall we pass? Theoretically, `storage_policy` configuration may be different on different replicas. So simply passing `volume_index` will not work in all cases.

Reply:
I think you should pass min / max ttl for the part which will be downloaded
So requesting replica can decide where it should go according to its rules at the very beginging

Reply:
Please don't answer here